intel/geoip/database.go

```diff
@@ -13,106 +13,205 @@ import (
+func (ub *updateBroadcaster) NeedsUpdate() bool {
+	ub.rw.RLock()
+	defer ub.rw.RUnlock()
 
-	dbDoReload.Set()
-	return doReload()
+	return ub.db == nil || ub.db.file.UpgradeAvailable()
 }
 
-func prepDatabaseForUse() error {
-	dbInUse.Set()
-	return doReload()
+// ReplaceDatabase replaces (or initialy sets) the mmdb database.
+// It also notifies all waiters about the availability of the new
+// database.
+func (ub *updateBroadcaster) ReplaceDatabase(db *geoIPDB) {
+	ub.rw.Lock()
+	defer ub.rw.Unlock()
+
+	if ub.db != nil {
+		ub.db.Close()
+	}
+	ub.db = db
+	ub.notifyWaiters()
 }
 
-func doReload() error {
-	// reload if needed
-	if dbDoReload.SetToIf(true, false) {
-		closeDBs()
-		if err := openDBs(); err != nil {
-			// try again the next time
-			dbDoReload.SetTo(true)
-			return err
-		}
+// notifyWaiters notifies and removes all waiters. Must be called
+// with ub.rw locked.
+func (ub *updateBroadcaster) notifyWaiters() {
+	waiters := ub.waiters
+	ub.waiters = nil
+	for _, c := range waiters {
+		close(c)
 	}
+}
 
-	return nil
+// getWaiter appends and returns a new waiter channel that gets closed
+// when a new database version is available. Must be called with
+// ub.rw locked.
+func (ub *updateBroadcaster) getWaiter() chan struct{} {
+	ch := make(chan struct{})
+	ub.waiters = append(ub.waiters, ch)
+	return ch
 }
 
-func openDBs() error {
-	var err error
+type updateWorker struct {
+	trigger chan struct{}
+	once    sync.Once
 
-	geoDBv4File, err = updates.GetFile("intel/geoip/geoipv4.mmdb.gz")
-	if err != nil {
-		return fmt.Errorf("could not get GeoIP v4 database file: %s", err)
+	v4 updateBroadcaster
+	v6 updateBroadcaster
+}
+
+// GetReader returns a MMDB reader for either the IPv4 or the IPv6 database.
+// If wait is true GetReader will wait at most 1 second for the database to
+// become available. If no database is available or GetReader times-out while
+// waiting nil is returned.
+func (upd *updateWorker) GetReader(v6 bool, wait bool) *maxminddb.Reader {
+	// check which updateBroadcaster we need to use
+	var ub *updateBroadcaster = &upd.v4
+	if v6 {
+		ub = &upd.v6
 	}
-	unpackedV4, err := geoDBv4File.Unpack(".gz", updater.UnpackGZIP)
-	if err != nil {
-		return err
+
+	// lock the updateBroadcaster and - if we are allowed to wait -
+	// create a new waiter channel, trigger an update and wait for at
+	// least 1 second for the update to complete.
+	ub.rw.Lock()
+	if ub.db == nil {
+		if wait {
+			waiter := ub.getWaiter()
+			ub.rw.Unlock()
+
+			upd.triggerUpdate()
+
+			select {
+			case <-waiter:
+				// call this method again but this time we don't allow
+				// it to wait since there must be a open database anyway ...
+				return upd.GetReader(v6, false)
+			case <-time.After(time.Second):
+				// we tried hard but failed so give up here
+				return nil
+			}
+		}
+		ub.rw.Unlock()
+		return nil
 	}
-	geoDBv4Reader, err = maxminddb.Open(unpackedV4)
-	if err != nil {
-		return err
+	rd := ub.db.Reader
+	ub.rw.Unlock()
```

Comment:
I'd really like to have the unlocks in a defer, so we don't deadlock ourselves.

Is a getter method for `ub.db` and locking for `getWaiter` a possibility?

Reply:
Not here because if I unlock a pending call to `ub.ReplaceDatabase()` might occur that we would miss then. I also like the defered ones more but I need to keep the lock locked :)

Reply:
Not exactly beautiful, but you could do the waiting itself in a defer that you define before the unlock defer:

Use a named return value of `r` and just return nil with `wait = true`:
```go
    defer func() {
        if wait {
	    select {
	    case <-waiter:
		    // call this method again but this time we don't allow
		    // it to wait since there must be a open database anyway ...
		    r = upd.GetReader(v6, false)
	    case <-time.After(time.Second):
		    // we tried hard but failed so give up here
	    }
	}
}()
```

Reply:
As the code in question currently cannot even panic, let's leave it as it is.